File: python/ndsq/src/ndsq/domain.py
"""The canonical input domain and its construction from raw JSON fields."""

from __future__ import annotations

from dataclasses import dataclass

from .errors import NdsqError, Reason
from .values import ImplicitValue, require_int, require_list, require_str_list


@dataclass
class Domain:
    """Per-dimension [inclusive_min, exclusive_max) plus labels."""

    rank: int
    inclusive_min: list[ImplicitValue]
    exclusive_max: list[ImplicitValue]
    labels: list[str]

    def to_json_fields(self) -> dict:
        return {
            "input_rank": self.rank,
            "input_inclusive_min": [v.to_json() for v in self.inclusive_min],
            "input_exclusive_max": [v.to_json() for v in self.exclusive_max],
            "input_labels": list(self.labels),
        }


def _bump_inclusive_to_exclusive(v: ImplicitValue) -> ImplicitValue:
    value = v.value + 1 if isinstance(v.value, int) else v.value
    return ImplicitValue(value=value, implicit=v.implicit)


def _add_shape(lo: ImplicitValue, sz: ImplicitValue) -> ImplicitValue:
    if isinstance(lo.value, int) and isinstance(sz.value, int):
        value: object = lo.value + sz.value
    elif lo.value == "+inf" or sz.value == "+inf":
        value = "+inf"
    else:  # any remaining -inf operand
        value = "-inf"
    return ImplicitValue(value=value, implicit=lo.implicit)  # type: ignore[arg-type]


def canonicalize_domain(
    *,
    rank: int | None = None,
    inclusive_min: list | None = None,
    exclusive_max: list | None = None,
    inclusive_max: list | None = None,
    shape: list | None = None,
    labels: list | None = None,
) -> Domain:
    """Reduce raw JSON-level domain fields to a canonical Domain."""

    def _bounds(raw: object | None, name: str) -> list[ImplicitValue] | None:
        if raw is None:
            return None
        return [ImplicitValue.from_json(b) for b in require_list(raw, name)]

    imin = _bounds(inclusive_min, "inclusive_min")
    emax = _bounds(exclusive_max, "exclusive_max")
    incmax = _bounds(inclusive_max, "inclusive_max")
    shp = _bounds(shape, "shape")
    labels = require_str_list(labels, "labels") if labels is not None else None
    if rank is not None:
        rank = require_int(rank, "rank")
        if rank < 0:
            raise NdsqError(Reason.INVALID_JSON, f"rank must be non-negative, got {rank}")

    upper_count = (emax is not None) + (incmax is not None) + (shp is not None)
    if upper_count > 1:
        raise NdsqError(
            Reason.MULTIPLE_UPPER_BOUNDS,
            "specify only one of exclusive_max / inclusive_max / shape",
        )

    resolved_rank = rank
    for arr in (imin, emax, incmax, shp, labels):
        if arr is None:
            continue
        if resolved_rank is not None and resolved_rank != len(arr):
            raise NdsqError(
                Reason.RANK_MISMATCH,
                f"array length {len(arr)} disagrees with rank {resolved_rank}",
            )
        resolved_rank = len(arr)
    r = resolved_rank if resolved_rank is not None else 0

    if imin is None:
        imin = [ImplicitValue.explicit(0) for _ in range(r)]

    if emax is not None:
        exclusive = emax
    elif incmax is not None:
        exclusive = [_bump_inclusive_to_exclusive(v) for v in incmax]
    elif shp is not None:
        exclusive = [_add_shape(lo, sz) for lo, sz in zip(imin, shp)]
    else:
        exclusive = [ImplicitValue(value="+inf", implicit=True) for _ in range(r)]

    out_labels = labels if labels is not None else ["" for _ in range(r)]
    return Domain(rank=r, inclusive_min=imin, exclusive_max=exclusive, labels=out_labels)

File: typescript/src/domain.ts
import { NdsqError, Reason } from "./errors.ts";
import {
  type BoundJson,
  type IndexValue,
  type ParsedBound,
  boundToJSON,
  demote,
  parseBound,
  requireArray,
  requireInt,
  requireStringArray,
  toBig,
} from "./values.ts";

export interface DomainFields {
  input_rank: number;
  input_inclusive_min: BoundJson[];
  input_exclusive_max: BoundJson[];
  input_labels: string[];
}

function bumpInclusiveToExclusive(b: ParsedBound): ParsedBound {
  const v = b.value;
  const value: IndexValue = v === "-inf" || v === "+inf" ? v : demote(toBig(v) + 1n);
  return { value, implicit: b.implicit };
}

function addShape(lo: ParsedBound, sz: ParsedBound): ParsedBound {
  const a = lo.value;
  const b = sz.value;
  let value: IndexValue;
  if (a === "-inf" || a === "+inf" || b === "-inf" || b === "+inf") {
    value = a === "+inf" || b === "+inf" ? "+inf" : "-inf";
  } else {
    value = demote(toBig(a) + toBig(b));
  }
  return { value, implicit: lo.implicit };
}

export interface RawDomainFields {
  rank?: unknown;
  inclusive_min?: unknown;
  exclusive_max?: unknown;
  inclusive_max?: unknown;
  shape?: unknown;
  labels?: unknown;
}

export function canonicalizeDomain(fields: RawDomainFields): DomainFields {
  const parseBounds = (raw: unknown, name: string): ParsedBound[] | null =>
    raw === undefined ? null : requireArray(raw, name).map((b) => parseBound(b));

  const imin = parseBounds(fields.inclusive_min, "inclusive_min");
  const emax = parseBounds(fields.exclusive_max, "exclusive_max");
  const incmax = parseBounds(fields.inclusive_max, "inclusive_max");
  const shp = parseBounds(fields.shape, "shape");
  const labels = fields.labels === undefined ? null : requireStringArray(fields.labels, "labels");
  // A rank is a small dimension count; reduce to number for length comparisons.
  let rank = fields.rank === undefined ? null : Number(requireInt(fields.rank, "rank"));
  if (rank !== null && rank < 0) {
    throw new NdsqError(Reason.InvalidJson, `rank must be non-negative, got ${rank}`);
  }

  const upperCount = (emax !== null ? 1 : 0) + (incmax !== null ? 1 : 0) + (shp !== null ? 1 : 0);
  if (upperCount > 1) {
    throw new NdsqError(Reason.MultipleUpperBounds, "specify only one of exclusive_max / inclusive_max / shape");
  }

  let resolved = rank;
  for (const arr of [imin, emax, incmax, shp, labels]) {
    if (arr === null) continue;
    if (resolved !== null && resolved !== arr.length) {
      throw new NdsqError(Reason.RankMismatch, `array length ${arr.length} disagrees with rank ${resolved}`);
    }
    resolved = arr.length;
  }
  const r = resolved ?? 0;

  const iminFinal: ParsedBound[] =
    imin ?? Array.from({ length: r }, () => ({ value: 0, implicit: false }) as ParsedBound);

  let exclusive: ParsedBound[];
  if (emax !== null) exclusive = emax;
  else if (incmax !== null) exclusive = incmax.map(bumpInclusiveToExclusive);
  else if (shp !== null) exclusive = iminFinal.map((lo, i) => addShape(lo, shp[i]));
  else exclusive = Array.from({ length: r }, () => ({ value: "+inf", implicit: true }) as ParsedBound);

  return {
    input_rank: r,
    input_inclusive_min: iminFinal.map(boundToJSON),
    input_exclusive_max: exclusive.map(boundToJSON),
    input_labels: labels ?? Array.from({ length: r }, () => ""),
  };
}
